### ISSUE TITLE
Write tests using Chrome's virtual authenticator API

### ISSUE THREAD

**Issue by lgarron** [github/webauthn-json#31]
I've held off on writing thorough tests, because getting a good mock WebAuthn authenticator working was a bit daunting. However, Chrome's DevTools protocol now supports virtual authenticators: https://chromedevtools.github.io/devtools-protocol/tot/WebAuthn/

This should make it possible to write fairly robust tests.

I haven't set up a GitHub action testing using the DevTools protocol, so we could use some help if anyone is familiar with how to set that up.

**Comment by halo** [comment]
Using Capybara and the [webauthn](https://github.com/cedarcode/webauthn-ruby) gem, I was able to get integration tests working in a Rails environment. I thought I'd post it here for a kickstart.

To my surprise, recent versions of selenium/webdrivers support this out-of-the-box.

```ruby
# Setup is only possible *after* an initial request (visit in Capybara) has been made.
visit '/login'

# Ensure same-origin
WebAuthn.configuration.origin = Capybara.current_session.server.base_url

# Enable virtual authenticators in browser
devtools = page.driver.browser.devtools
devtools.send_cmd 'WebAuthn.enable'

# Create an Authenticator
# See https://chromedevtools.github.io/devtools-protocol/tot/WebAuthn/#type-VirtualAuthenticatorOptions
options = {
  protocol: :ctap2,
  transport: :internal,
  hasResidentKey: false, # Chrome should not have to reveal a list of existing virtual authenticator IDs.
  # isUserConsenting: true, # Not sure, this option exists in selenium but not in chrome?
  hasUserVerification: has_user_verification,
  isUserVerified: true,
}
attributes = record.devtools.send_cmd 'WebAuthn.addVirtualAuthenticator', options: options
id = attributes.dig('result', 'authenticatorId')

# Make sure to tear down after each test because it will interfere with further tests
devtools.send_cmd 'WebAuthn.removeVirtualAuthenticator', authenticatorId: id
devtools.send_cmd 'WebAuthn.disable'
```

**Comment by halo** [comment]
Thanks to [this commit](https://github.com/SeleniumHQ/selenium/pull/10903/files#diff-521c79fc23030ae5f66dd1e4e5d140af70f6be7becd55f1cd1989057954020d4) it just got much simpler:

```ruby
visit '/login'

options = ::Selenium::WebDriver::VirtualAuthenticatorOptions.new
page.driver.browser.add_virtual_authenticator(options)

click_on 'Start the registration ceremony using my new hardware key'
```

**Comment by lgarron** [comment]
> Thanks to [this commit](https://github.com/SeleniumHQ/selenium/pull/10903/files#diff-521c79fc23030ae5f66dd1e4e5d140af70f6be7becd55f1cd1989057954020d4) it just got much simpler:

Thanks!

Just to be transparent, though: In the interest of keeping this library as easy as possible to maintain for its limited remaining lifetime, I think any JS testing framework would be preferable over Ruby (or over any other separate language).

**Comment by halo** [comment]
I completely agree, thanks for being clear. I just wanted to post this as a heads-up that selenium is in the progress of directly implementing those DevTools calls. Three months ago [JS got them as well](https://github.com/SeleniumHQ/selenium/pull/10663/files). Good luck!

**Comment by fletchto99** [comment]
Closing as this project has been deprecated. Please see the project readme for more information.
